File: 25_reverse_nodes_in_k_group/main.py
# -*- coding: utf-8 -*-


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):

    def reverse_k_group(self, head, k):
        p, c = head, 0
        while p is not None and c < k:
            p, c = p.next, c + 1
        if c == k:
            p = self.reverse_k_group(p, k)
            while c > 0:
                tmp = head.next
                head.next = p
                p = head
                head = tmp
                c -= 1
            head = p
        return head


def output(l):
    p = l
    while p is not None:
        print(p.val),
        p = p.next
    print('')
    

if __name__ == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node1.next = node2
    node3 = ListNode(3)
    node2.next = node3
    node4 = ListNode(4)
    node3.next = node4
    node5 = ListNode(5)
    node4.next = node5
    output(node1)

    s = Solution()
    node1 = s.reverse_k_group(node1, 2)
    output(node1)

File: 25_reverse_nodes_in_k_group/go/main.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	p, c := head, 0
	for p != nil && c < k {
		p, c = p.Next, c + 1
	}
	if c == k {
		p = reverseKGroup(p, k)
		for ; c > 0; c-- {
			tmp := head.Next
			head.Next = p
			p = head
			head = tmp
		}
		head = p
	}
	return head
}

func output(l *ListNode) {
	for p := l; p != nil; p = p.Next {
		fmt.Printf("%d ", p.Val)
	}
	fmt.Println()
}

func main() {
	node1 := &ListNode{Val: 1}
	node2 := &ListNode{Val: 2}
	node1.Next = node2
	node3 := &ListNode{Val: 3}
	node2.Next = node3
	node4 := &ListNode{Val: 4}
	node3.Next = node4
	node5 := &ListNode{Val: 5}
	node4.Next = node5
	output(node1)

	node1 = reverseKGroup(node1, 2)
	output(node1)
}
